File: src/services/storageService.ts
import sharp from 'sharp';
import fs from 'fs/promises';
import { unlink } from 'fs';

interface ProcessedImage {
  data: Buffer;
  info: sharp.OutputInfo;
}

/**
 * Service that reads and returns an image suitable for an OCR task
 * @param imagePath - The path where to read the image
 * @returns Promise<ProcessedImage> The processed image for OCR
 * @throws Error if image processing fails or file doesn't exist
 */
async function getOcrImage(imagePath: string): Promise<ProcessedImage> {
  try {
    // Verify file exists before processing
    await fs.access(imagePath);
    
    const image = sharp(imagePath);
    const metadata = await image.metadata();

    if (!metadata) {
      throw new Error('Failed to read image metadata');
    }

    // If image has alpha channel (transparency)
    if (metadata.hasAlpha) {
      return await image
        .removeAlpha()
        .ensureAlpha()
        .composite([
          {
            input: Buffer.from([255, 255, 255, 255]),
            raw: {
              width: 1,
              height: 1,
              channels: 4
            },
            tile: true,
            blend: 'dest-over'
          }
        ])
        .toColorspace('srgb')
        .toBuffer({ resolveWithObject: true });
    }

    // If image doesn't have alpha channel
    return await image
      .toColorspace('srgb')
      .toBuffer({ resolveWithObject: true });
  } catch (error: unknown) {
    console.error('Error processing OCR image:', error);
    if (error instanceof Error) {
      throw new Error(`Failed to process OCR image: ${error.message}`);
    }
    throw new Error('Failed to process OCR image: Unknown error');
  }
}

/**
 * Service that reads and returns an image suitable for Object/KeyPoints detection
 * @param imagePath - The path where to read the image
 * @returns Promise<ProcessedImage> The processed image for prediction
 * @throws Error if image processing fails or file doesn't exist
 */
async function getPredictImage(imagePath: string): Promise<ProcessedImage> {
  try {
    // Verify file exists before processing
    await fs.access(imagePath);
    
    const image = sharp(imagePath);
    return await image
      .toColorspace('srgb')
      .toBuffer({ resolveWithObject: true });
  } catch (error) {
    console.error('Error processing predict image:', error);
    if (error instanceof Error) {
      throw new Error(`Failed to process predict image: ${error.message}`);
    }
    throw new Error('Failed to process predict image: Unknown error');
  }
}

/**
 * Service that returns the images for OCR and Object/KeyPoints detection tasks
 * @param path - The local path where the image is downloaded
 * @returns Promise<[ProcessedImage, ProcessedImage]> Tuple of processed images for OCR and predictions
 * @throws Error if image processing fails or file operations fail
 */
async function getOcrAndPredictImages(
  path: string
): Promise<[ProcessedImage, ProcessedImage]> {
  try {
    // Verify file exists before processing
    await fs.access(path);

    const [ocrImg, predictImg] = await Promise.all([
      getOcrImage(path),
      getPredictImage(path)
    ]);

    try {
      // Remove temporary file using fs.promises instead of callback
      await fs.unlink(path);
    } catch (unlinkError) {
      console.warn(`Failed to remove temporary file ${path}:`, unlinkError);
      // Continue execution even if file deletion fails
    }

    return [ocrImg, predictImg];
  } catch (error) {
    console.error('Error processing images:', error);
    if (error instanceof Error) {
      throw new Error(`Failed to process images: ${error.message}`);
    }
    throw new Error('Failed to process images: Unknown error');
  }
}

export const storageService = {
  getOcrImage,
  getPredictImage,
  getOcrAndPredictImages,
};

File: src/i2b-backend/bpmn_redrawer_backend/api/services/storage_service.py
import numpy as np
import cv2 as cv2

def get_ocr_image(image: np.ndarray):
    """Service that reads and returns an image suitable for an OCR task, given its ndarray

    Parameters
    ----------
    image_path: str
        The path where to read the image

    Returns
    -------
    ndarray
        The image for the Object/KeyPoints detection
    """

    if image is not None and image.shape[2] == 4:
        trans_mask = image[:, :, 3] == 0
        image[trans_mask] = [255, 255, 255, 255]
        image = (
            image.astype(np.uint16)
            + 255
            - np.repeat(np.expand_dims(image[:, :, 3], 2), 4, axis=2)
        )
        image = np.ndarray.clip(image, 0, 255)
        image = image[:, :, [0, 1, 2]]
        image = np.ascontiguousarray(image, dtype=np.uint8)
    return image

def get_ocr_and_predict_images(image_bytes: bytes):
    """Service that returns the images for the OCR and Object/KeyPoints detection tasks.

    Parameters
    ----------
    image_bytes: bytes
        user uploaded image bytes

    Returns
    -------
    tuple
        The two images for OCR and predictions

    """
    # Convert bytes to numpy array
    image_array = np.asarray(bytearray(image_bytes), dtype=np.uint8)
    
    # Decode the array to an image
    image = cv2.imdecode(image_array, cv2.IMREAD_COLOR)

    ocr_img = get_ocr_image(image)
    predict_img = image

    return ocr_img, predict_img
